File: chat_server/watercooler/web/channels/chat_room.ex
defmodule Watercooler.ChatRoom do
  use Phoenix.Channel

  alias Watercooler.Domain.ChatServer, as: Server
  alias Watercooler.Domain.ChatRoom,   as: Room

  def join(socket, "chat_room", _message) do
    {:ok, socket}
  end

  def join(socket, _no, _message) do
    {:error, socket, :unauthorized}
  end

  @room_name "The Watercooler"

  def event(socket, "chat_room:join", payload = %{"nickname" => nickname}) do
    Server.create_room(@room_name)
    {:ok, room} = Server.room(@room_name)
    Room.join(room, nickname)
    broadcast socket, "chat_room:joined", %{nickname: nickname, users: Room.users(room)}
    socket
  end

  def event(socket, "chat_room:leave", payload = %{"nickname" => nickname}) do
    {:ok, room} = Server.room(@room_name)
    Room.leave(room, nickname)
    broadcast socket, "chat_room:left", %{nickname: nickname, users: Room.users(room)}
    socket
  end

  def event(socket, "chat_room:message", payload = %{"nickname" => _nickname, "message" => _message}) do
    broadcast socket, "chat_room:message", payload
    socket
  end

end

File: chat_server/watercooler/priv/static/js/chat_room.js
$(function() {

  var ChatRoom = {
    people: [],
    nickname: "Troubadour2000",
    ui: {
      refreshRoster: function () {
        $('#roster-count').html(ChatRoom.people.length)
        $('#roster').html(ChatRoom.people)
      },
      appendMessage: function (nickname, msg) {
        var html = '[' + nickname + '] ' + msg + '<br>';
        $('#messages').append(html);
      }
    }
  }

  var socket = new Phoenix.Socket("/ws");

  socket.join("chat_room", "chat_room", {foo: "bar"}, function(channel) {

    channel.send("chat_room:join", {nickname: ChatRoom.nickname});

    channel.on("chat_room:joined", function(message) {
      ChatRoom.people.push(message.nickname)
      ChatRoom.ui.appendMessage(message.nickname, "*** Entered the room ***")
      ChatRoom.ui.refreshRoster();
    });

    channel.on("chat_room:left", function(message) {
      ChatRoom.ui.appendMessage(message.nickname, "*** Left the room ***")
      ChatRoom.ui.refreshRoster();
    });

    channel.on("chat_room:message", function(message) {
      ChatRoom.ui.appendMessage(message.nickname, message.message)
    });

    $('#leave-room-button').click(function() {
      channel.send("chat_room:leave", {nickname: ChatRoom.nickname});
    });

    sendMessage = function () {
      var message = $('#message').val();
      channel.send("chat_room:message", {nickname: ChatRoom.nickname, message: message});
      $('#message').val('');
    }

    $('#message').keypress(function (event) {
      if (event.charCode == 13) {
        sendMessage()
      }
    });

    $('#send-message-button').click(sendMessage)

  });

});